pkg/storage/ingest/reader.go

```diff
@@ -0,0 +1,234 @@
+	"github.com/grafana/dskit/services"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/twmb/franz-go/pkg/kadm"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/plugin/kprom"
+)
+
+// consumerGroup is only used to store commit offsets, not for actual consuming.
+const consumerGroup = "mimir"
+
+type Record struct {
+	TenantID string
+	Content  []byte
+}
+
+type RecordConsumer interface {
+	Consume(context.Context, Record) error
+}
+
+type PartitionReader struct {
+	services.Service
+
+	kafkaAddress string
+	kafkaTopic   string
+	partition    int32
+
+	client *kgo.Client
+
+	consumer RecordConsumer
+	metrics  *readerMetrics
+
+	logger log.Logger
+	reg    prometheus.Registerer
+}
+
+func NewReader(kafkaAddress, kafkaTopic string, partitionID int32, consumer RecordConsumer, logger log.Logger, reg prometheus.Registerer) (*PartitionReader, error) {
+	metrics := newReaderMetrics(reg)
+	r := &PartitionReader{
+		kafkaAddress: kafkaAddress,
+		kafkaTopic:   kafkaTopic,
+		partition:    partitionID,
+		reg:          reg,
+		consumer:     consumer, // TODO consume records in parallel
+		metrics:      metrics,
+		logger:       log.With(logger, "partition", partitionID),
+	}
+
+	r.Service = services.NewBasicService(r.start, r.run, r.stop)
+	return r, nil
+}
+
+func (r *PartitionReader) start(ctx context.Context) error {
+	offset, err := r.fetchLastCommittedOffset(ctx)
+	if err != nil {
+		return err
+	}
+	level.Info(r.logger).Log("msg", "resuming consumption from offset", "offset", offset)
+
+	r.client, err = r.newKafkaReader(offset, r.reg)
+	if err != nil {
+		return errors.Wrap(err, "creating kafka reader client")
+	}
+	return nil
+}
+
+func (r *PartitionReader) run(ctx context.Context) error {
+	consumeCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for ctx.Err() == nil {
+		fetches := r.client.PollFetches(ctx)
+		if fetches.Err() != nil {
+			if errors.Is(fetches.Err(), context.Canceled) {
+				return nil
+			}
+			err := collectFetchErrs(fetches)
+			level.Error(r.logger).Log("msg", "encountered error while fetching", "err", err)
+			continue
+		}
+		level.Debug(r.logger).Log("msg", "fetched records", "num_records", fetches.NumRecords())
+
+		r.recordFetchesLag(fetches)
+		r.consumeFetches(consumeCtx, fetches)
+		r.commitFetches(consumeCtx, fetches)
+	}
+
+	return nil
+}
+
+func collectFetchErrs(fetches kgo.Fetches) (_ error) {
+	mErr := multierror.New()
+	fetches.EachError(func(s string, i int32, err error) {
+		// TODO handle errors properly, there can be some error we can ignore and some errors for which we have to reset the kafka client. See docs on EachError
+		mErr.Add(err)
+	})
+	return mErr.Err()
+}
+
+func (r *PartitionReader) commitFetches(ctx context.Context, fetches kgo.Fetches) {
+	committed, err := kadm.NewClient(r.client).CommitOffsets(ctx, consumerGroup, kadm.OffsetsFromFetches(fetches))
```

Comment:
Can we reuse the `kadm.Client` here?

Reply:
Note: we're not closing the `kadm.Client` but that looks fine because if we close it would also close `r.client`.